File: src/catchido/scrapers/tiktok.py
import re
import json
import asyncio
import random
from typing import List, Optional, Dict, Any
from loguru import logger
import httpx
from bs4 import BeautifulSoup

from .base import BaseScraper
from ..db.models import MediaItem, MediaType

class TikTokScraper(BaseScraper):
    def __init__(self, session_cookie: str, config):
        super().__init__(config, request_delay=config.tiktok.request_delay)
        self.session_cookie = session_cookie

    def get_hd_url(self, url: str) -> str:
        """
        TikTok photo URLs are usually HD.
        We can strip query parameters or modify image size parameters if needed.
        """
        return url

    async def _create_session(self) -> httpx.AsyncClient:
        client = await super()._create_session()
        client.headers.update({
            "Referer": "https://www.tiktok.com/",
            "Sec-Fetch-Dest": "document",
            "Sec-Fetch-Mode": "navigate",
            "Sec-Fetch-Site": "none"
        })
        if self.session_cookie:
            # TikTok session cookie is usually called 'sessionid' or similar,
            # we will set the Cookie header directly.
            client.headers["Cookie"] = f"sessionid={self.session_cookie}"
            logger.debug("TikTok session cookie loaded")
        return client

    async def fetch_media(
        self, 
        query_or_username: str, 
        since_id: Optional[str] = None, 
        limit: Optional[int] = None
    ) -> List[MediaItem]:
        """Fetch media items (photo slides) from a TikTok user profile."""
        username = query_or_username.lstrip("@")
        if not self.client:
            await self._create_session()

        url = f"https://www.tiktok.com/@{username}"
        logger.info("Fetching TikTok profile for @{}", username)

        try:
            await self._throttle()
            resp = await self.client.get(url)
            if resp.status_code != 200:
                logger.error("TikTok profile fetch failed for @{} (Status {}).", username, resp.status_code)
                return []

            html = resp.text
            media_items = self._parse_tiktok_html(html, username)
            
            filtered_items = []
            for item in media_items:
                if since_id and item.post_id == since_id:
                    logger.info("Reached TikTok checkpoint since_id: {}. Stopping.", since_id)
                    break
                filtered_items.append(item)
                if limit and len(filtered_items) >= limit:
                    break

            logger.info("Retrieved {} media items for @{} from TikTok profile.", len(filtered_items), username)
            return filtered_items
        except Exception as e:
            logger.exception("Error during TikTok scraping for @{}: {}", username, e)
            return []

    async def fetch_media_from_url(self, url: str) -> List[MediaItem]:
        """Fetch media items directly from a specific TikTok post URL."""
        if not self.client:
            await self._create_session()

        logger.info("Fetching TikTok post from URL: {}", url)
        try:
            await self._throttle()
            # Handle shortened links (e.g. vm.tiktok.com / vt.tiktok.com)
            resp = await self.client.get(url, follow_redirects=True)
            if resp.status_code != 200:
                logger.error("Failed to fetch TikTok post URL (Status {}).", resp.status_code)
                return []

            # Extract post owner/author from redirected URL or HTML
            author = "unknown"
            match = re.search(r'tiktok\.com/@([^/]+)', str(resp.url))
            if match:
                author = match.group(1)

            html = resp.text
            return self._parse_tiktok_html(html, author)
        except Exception as e:
            logger.exception("Error fetching TikTok post URL {}: {}", url, e)
            return []

    def _parse_tiktok_html(self, html: str, author: str) -> List[MediaItem]:
        """Parse raw TikTok HTML and extract photo slides from state JSON script tags."""
        media_items = []
        soup = BeautifulSoup(html, "lxml")

        # Find rehydration or state script tag
        script_tags = [
            soup.find("script", id="__UNIVERSAL_DATA_FOR_REHYDRATION__"),
            soup.find("script", id="SIGI_STATE"),
            soup.find("script", id="sigi-state")
        ]
        
        # Also look in all script tags containing typical TikTok JSON
        for tag in soup.find_all("script"):
            if tag.string and ("__UNIVERSAL_DATA_FOR_REHYDRATION__" in tag.string or "SIGI_STATE" in tag.string or "itemStruct" in tag.string):
                script_tags.append(tag)

        for tag in script_tags:
            if not tag or not tag.string:
                continue
            
            content = tag.string.strip()
            # Clean up JS assignment if needed (e.g. window.SIGI_STATE = ...)
            if "=" in content and not content.startswith("{"):
                match = re.search(r'=\s*(\{.*?\});?$', content, re.DOTALL)
                if match:
                    content = match.group(1)

            try:
                data = json.loads(content)
                items = self._extract_media_from_json(data, author)
                media_items.extend(items)
            except Exception:
                pass

        # Fallback: Regex parse TikTok photo CDN urls if JSON failed
        if not media_items and self.config.general.download_photos:
            logger.debug("Falling back to regex parsing for TikTok photo CDN URLs")
            # TikTok CDNs use tiktokcdn.com / byteoversea.com / ibyteimg.com / tiktokcdn-us.com
            cdn_pattern = r'https://[a-zA-Z0-9.-]+\.(?:tiktokcdn|byteoversea|ibyteimg|tiktokcdn-us)\.com/[^"\']+\.(?:jpg|jpeg|png|webp)'
            urls = re.findall(cdn_pattern, html)
            unique_urls = list(set(urls))
            
            for idx, cdn_url in enumerate(unique_urls):
                clean_url = cdn_url.replace("\\u0026", "&").replace("&amp;", "&")
                import hashlib
                post_id = hashlib.md5(clean_url.encode('utf-8')).hexdigest()[:16]

                media_items.append(MediaItem(
                    url=self.get_hd_url(clean_url),
                    platform="tiktok",
                    post_id=post_id,
                    author=author,
                    media_type=MediaType.IMAGE,
                    original_url=clean_url,
                    text=""
                ))

        # De-duplicate items list
        unique_results = {}
        for item in media_items:
            unique_results[item.url] = item
        return list(unique_results.values())

    def _extract_media_from_json(self, data: Any, author: str) -> List[MediaItem]:
        """Recursively scan JSON tree for TikTok post structs and map to MediaItem."""
        items = []

        if isinstance(data, dict):
            # Check for itemStruct or individual item representation
            is_post = "id" in data and ("imagePost" in data or "video" in data or "desc" in data)
            if is_post:
                post_id = data.get("id") or str(random.randint(100000, 999999))
                caption_text = data.get("desc") or data.get("caption") or ""
                hashtags = [tag.strip("#") for tag in re.findall(r'#\w+', caption_text)]
                
                created_at_ts = data.get("createTime") or data.get("addedTime")
                created_at = None
                if created_at_ts:
                    from datetime import datetime
                    created_at = datetime.utcfromtimestamp(int(created_at_ts)).isoformat()

                # Process photo slideshow (imagePost)
                image_post_info = data.get("imagePost") or data.get("image_post_info")
                if image_post_info and self.config.general.download_photos:
                    images = image_post_info.get("images", [])
                    for idx, img in enumerate(images):
                        # Extract URL candidates
                        display_url = None
                        # structure can be img['display_image']['url_list'][0] or similar
                        url_list = img.get("display_image", {}).get("url_list", []) or img.get("url_list", []) or []
                        if url_list:
                            display_url = url_list[0]
                        else:
                            display_url = img.get("url")
                            
                        width = img.get("width") or img.get("display_image", {}).get("width")
                        height = img.get("height") or img.get("display_image", {}).get("height")
                        
                        if display_url:
                            items.append(MediaItem(
                                url=self.get_hd_url(display_url),
                                platform="tiktok",
                                post_id=f"{post_id}_{idx}",
                                author=author,
                                media_type=MediaType.IMAGE,
                                width=width,
                                height=height,
                                original_url=display_url,
                                created_at=created_at,
                                text=caption_text,
                                hashtags=hashtags
                            ))
                
                # Check for video if download_videos is enabled
                elif "video" in data and self.config.general.download_videos:
                    video_struct = data.get("video", {})
                    # Get downloadAddr or playAddr
                    url = video_struct.get("downloadAddr") or video_struct.get("playAddr")
                    width = video_struct.get("width")
                    height = video_struct.get("height")
                    if url:
                        items.append(MediaItem(
                            url=url,
                            platform="tiktok",
                            post_id=post_id,
                            author=author,
                            media_type=MediaType.VIDEO,
                            width=width,
                            height=height,
                            original_url=url,
                            created_at=created_at,
                            text=caption_text,
                            hashtags=hashtags
                        ))
            else:
                for val in data.values():
                    items.extend(self._extract_media_from_json(val, author))
        elif isinstance(data, list):
            for val in data:
                items.extend(self._extract_media_from_json(val, author))

        return items

File: src-tauri/src/scrapers/tiktok.rs
use regex::Regex;
use serde_json::Value;
use std::time::Duration;

use crate::config::AppConfig;
use super::MediaItem;

pub struct TikTokScraper {
    client: reqwest::Client,
    session_cookie: String,
    download_photos: bool,
    download_videos: bool,
}

impl TikTokScraper {
    pub fn new(session_cookie: String, config: &AppConfig) -> Self {
        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert("Referer", reqwest::header::HeaderValue::from_static("https://www.tiktok.com/"));
        headers.insert("Sec-Fetch-Dest", reqwest::header::HeaderValue::from_static("document"));
        headers.insert("Sec-Fetch-Mode", reqwest::header::HeaderValue::from_static("navigate"));
        headers.insert("Sec-Fetch-Site", reqwest::header::HeaderValue::from_static("none"));
        headers.insert(
            "User-Agent",
            reqwest::header::HeaderValue::from_static(
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            ),
        );
        
        if !session_cookie.is_empty() {
            let cookie_str = format!("sessionid={}", session_cookie);
            if let Ok(cookie_val) = reqwest::header::HeaderValue::from_str(&cookie_str) {
                headers.insert("Cookie", cookie_val);
            }
        }

        let client = reqwest::Client::builder()
            .default_headers(headers)
            .timeout(Duration::from_secs(15))
            .build()
            .unwrap_or_default();

        Self {
            client,
            session_cookie,
            download_photos: config.general.download_photos,
            download_videos: config.general.download_videos,
        }
    }

    pub fn get_hd_url(&self, url: &str) -> String {
        url.to_string()
    }

    pub async fn fetch_media(
        &self,
        query_or_username: &str,
        since_id: Option<&str>,
        limit: Option<usize>,
    ) -> Result<Vec<MediaItem>, String> {
        let username = query_or_username.trim_start_matches('@');
        let url = format!("https://www.tiktok.com/@{}", username);
        
        let resp = self.client.get(&url).send().await
            .map_err(|e| format!("TikTok profile request failed: {}", e))?;
            
        if resp.status() != reqwest::StatusCode::OK {
            return Err(format!("TikTok profile returned status: {}", resp.status()));
        }

        let html = resp.text().await
            .map_err(|e| format!("Failed to read TikTok response: {}", e))?;
            
        let mut media_items = Vec::new();
        
        // 1. Find rehydration or state script tags
        let state_patterns = [
            r#"<script id="__UNIVERSAL_DATA_FOR_REHYDRATION__"[^>]*>([\s\S]*?)</script>"#,
            r#"<script id="SIGI_STATE"[^>]*>([\s\S]*?)</script>"#,
            r#"<script id="sigi-state"[^>]*>([\s\S]*?)</script>"#
        ];

        for pattern in &state_patterns {
            let re = Regex::new(pattern).unwrap();
            for cap in re.captures_iter(&html) {
                let mut content = cap.get(1).map(|m| m.as_str()).unwrap_or("").trim().to_string();
                
                // Clean up JS assignment if needed (e.g. window.SIGI_STATE = ...)
                if content.contains('=') && !content.starts_with('{') {
                    let assign_re = Regex::new(r"=\s*(\{.*?\});?$").unwrap();
                    if let Some(m) = assign_re.captures(&content) {
                        content = m.get(1).map(|x| x.as_str()).unwrap_or("").to_string();
                    }
                }
                
                if let Ok(json_val) = serde_json::from_str::<Value>(&content) {
                    let mut items = Vec::new();
                    self.extract_media_from_json(&json_val, username, &mut items);
                    media_items.extend(items);
                }
            }
        }

        // 2. Fallback: Regex parse TikTok photo CDN URLs
        if media_items.is_empty() && self.download_photos {
            log::debug!("Falling back to regex parsing for TikTok photo CDN URLs");
            let cdn_re = Regex::new(r#"https://[a-zA-Z0-9.-]+\.(?:tiktokcdn|byteoversea|ibyteimg|tiktokcdn-us)\.com/[^"']+\.(?:jpg|jpeg|png|webp)"#).unwrap();
            
            let mut unique_urls = Vec::new();
            for cap in cdn_re.captures_iter(&html) {
                let cdn_url = cap.get(0).map(|m| m.as_str()).unwrap_or("");
                if !unique_urls.contains(&cdn_url.to_string()) {
                    unique_urls.push(cdn_url.to_string());
                }
            }

            for cdn_url in unique_urls {
                let clean_url = cdn_url.replace("\\u0026", "&").replace("&amp;", "&");
                let hash = md5::compute(clean_url.as_bytes());
                let post_id = format!("{:x}", hash)[..16].to_string();

                media_items.push(MediaItem {
                    url: self.get_hd_url(&clean_url),
                    platform: "tiktok".to_string(),
                    post_id,
                    author: username.to_string(),
                    media_type: "photo".to_string(),
                    original_url: clean_url,
                    text: "".to_string(),
                });
            }
        }

        // Deduplicate
        let mut unique_results = std::collections::HashMap::new();
        for item in media_items {
            unique_results.insert(item.url.clone(), item);
        }
        
        let mut final_list: Vec<MediaItem> = unique_results.into_values().collect();
        
        // Filter by since_id
        if let Some(sid) = since_id {
            if let Some(pos) = final_list.iter().position(|item| item.post_id == sid) {
                log::info!("Reached TikTok checkpoint: {}. Stopping.", sid);
                final_list.truncate(pos);
            }
        }

        if let Some(lim) = limit {
            if final_list.len() > lim {
                final_list.truncate(lim);
            }
        }

        Ok(final_list)
    }

    fn extract_media_from_json(&self, data: &Value, author: &str, items: &mut Vec<MediaItem>) {
        if let Some(obj) = data.as_object() {
            let is_post = obj.contains_key("id") && 
                (obj.contains_key("imagePost") || obj.contains_key("video") || obj.contains_key("desc"));
                
            if is_post {
                let post_id = obj.get("id").and_then(|i| i.as_str()).unwrap_or("").to_string();
                let caption_text = obj.get("desc").or_else(|| obj.get("caption")).and_then(|t| t.as_str()).unwrap_or("").to_string();
                
                // Process photo slideshow
                if let Some(image_post_info) = obj.get("imagePost").or_else(|| obj.get("image_post_info")) {
                    if self.download_photos {
                        if let Some(images) = image_post_info.get("images").and_then(|i| i.as_array()) {
                            for (idx, img) in images.iter().enumerate() {
                                let mut display_url = None;
                                if let Some(url_list) = img.pointer("/display_image/url_list").or_else(|| img.get("url_list")).and_then(|l| l.as_array()) {
                                    if let Some(first_url) = url_list.get(0).and_then(|u| u.as_str()) {
                                        display_url = Some(first_url.to_string());
                                    }
                                } else if let Some(url) = img.get("url").and_then(|u| u.as_str()) {
                                    display_url = Some(url.to_string());
                                }
                                
                                if let Some(url) = display_url {
                                    items.push(MediaItem {
                                        url: self.get_hd_url(&url),
                                        platform: "tiktok".to_string(),
                                        post_id: format!("{}_{}", post_id, idx),
                                        author: author.to_string(),
                                        media_type: "photo".to_string(),
                                        original_url: url,
                                        text: caption_text.clone(),
                                    });
                                }
                            }
                        }
                    }
                }
                // Process video
                else if obj.contains_key("video") && self.download_videos {
                    let video_struct = obj.get("video").unwrap();
                    let url = video_struct.get("downloadAddr").or_else(|| video_struct.get("playAddr")).and_then(|u| u.as_str());
                    if let Some(video_url) = url {
                        items.push(MediaItem {
                            url: video_url.to_string(),
                            platform: "tiktok".to_string(),
                            post_id: post_id.clone(),
                            author: author.to_string(),
                            media_type: "video".to_string(),
                            original_url: video_url.to_string(),
                            text: caption_text.clone(),
                        });
                    }
                }
            } else {
                for value in obj.values() {
                    self.extract_media_from_json(value, author, items);
                }
            }
        } else if let Some(arr) = data.as_array() {
            for value in arr {
                self.extract_media_from_json(value, author, items);
            }
        }
    }
}
